### ISSUE TITLE
raspberry pi 4b 8G freebsd13 will crash

### ISSUE THREAD

**Issue by fundon** [libcala/whoami#20]
https://github.com/launchbadge/sqlx/issues/903

**Comment by AldaronLau** [comment]
@fundon Do you know if this crashes on FreeBSD with other architectures (non-ARM)?

**Comment by fundon** [comment]
Its ok on x86 platform.

**Comment by AldaronLau** [comment]
@fundon I do not have a Raspberry Pi 4, and I've never used FreeBSD - so I can't directly test it.  Is it a segmentation fault, panic on unwrap or something else?  A backtrace would be useful if it's a panic.  This could also maybe be a FreeBSD bug?  It looks like raspberry pi support is pretty new, and there's no stability guarantees for the only supported version (13).  Try installing the system `whoami` command (that exists on freebsd, right - something like `pkg install whoami`?), and see if that crashes.

**Comment by fundon** [comment]
```rust
fn main() {
    println!("WhoAmI {}", env!("CARGO_PKG_VERSION"));
    println!();
    println!(
        "User's Name            whoami::realname():    {}",
        whoami::realname()
    );
    let username: Option<String> = None;
    println!(
        "User's Username        whoami::username():    {}",
        username.unwrap_or_else(whoami::username)
        //whoami::username()
    );
    println!(
        "Device's Pretty Name   whoami::devicename():  {}",
        whoami::devicename()
    );
    println!(
        "Device's Hostname      whoami::hostname():    {}",
        whoami::hostname()
    );
    println!(
        "Device's Platform      whoami::platform():    {}",
        whoami::platform()
    );
    println!(
        "Device's OS Distro     whoami::distro():      {}",
        whoami::distro()
    );
    println!(
        "Device's Desktop Env.  whoami::desktop_env(): {}",
        whoami::desktop_env()
    );
}
```

### Debug
```
freebsd@rpi4g ~/c/whoami (master)> cargo r --example whoami-demo
    Finished dev [unoptimized + debuginfo] target(s) in 0.18s
     Running `target/debug/examples/whoami-demo`
WhoAmI 1.0.0

User's Name            whoami::realname():    Freebsd
[src/unix.rs:157] string_from_os(username_os()) = "freebsd"
User's Username        whoami::username():    freebsd
Device's Pretty Name   whoami::devicename():  Rpi4g
Device's Hostname      whoami::hostname():    rpi4g
Device's Platform      whoami::platform():    BSD
Device's OS Distro     whoami::distro():      FreeBSD 13.0-CURRENT
Device's Desktop Env.  whoami::desktop_env(): Unknown: Unknown
```

### Release
```
freebsd@rpi4g ~/c/whoami (master)> cargo r --release --example whoami-demo
    Finished release [optimized] target(s) in 0.07s
     Running `target/release/examples/whoami-demo`
WhoAmI 1.0.0

User's Name            whoami::realname():    Freebsd
[src/unix.rs:157] string_from_os(username_os()) = "freebsd"
fish: 'cargo r --release --example who…' terminated by signal SIGBUS (Misaligned address error)
```

Crash file
[whoami-demo.core.zip](https://github.com/libcala/whoami/files/5700561/whoami-demo.core.zip)

**Comment by fundon** [comment]
Maybe its Rust bug!

**Comment by AldaronLau** [comment]
@fundon Looks like it's at least possible it's a Rust bug (or some really obscure unsound code execution in whoami).  Quickly looking at the Rust issues I don't see anything related to this that's been reported yet.  I wonder how difficult it would be to make a minimized version of the code needed to produce this error (no dependencies).

**Comment by fundon** [comment]
```rust
fn main() {
    println!("WhoAmI {}", env!("CARGO_PKG_VERSION"));
    println!();
    println!(
        "User's Name            whoami::realname():    {}",
        whoami::realname()
    );
    println!(
        "User's Username        whoami::username():    {}",
        whoami::username()
    );
    println!(
        "Device's Pretty Name   whoami::devicename():  {}",
        whoami::devicename()
    );
    println!(
        "Device's Hostname      whoami::hostname():    {}",
        whoami::hostname()
    );
    println!(
        "Device's Platform      whoami::platform():    {}",
        whoami::platform()
    );
    println!(
        "Device's OS Distro     whoami::distro():      {}",
        whoami::distro()
    );
    println!(
        "Device's Desktop Env.  whoami::desktop_env(): {}",
        whoami::desktop_env()
    );
}
```

```shell
freebsd@rpi4g ~/c/whoami (master)> cargo r --example whoami-demo
    Finished dev [unoptimized + debuginfo] target(s) in 0.18s
     Running `target/debug/examples/whoami-demo`
WhoAmI 1.0.0

User's Name            whoami::realname():    Freebsd
[src/unix.rs:157] string_from_os(username_os()) = "freebsd"
User's Username        whoami::username():    freebsd
Device's Pretty Name   whoami::devicename():  Rpi4g
Device's Hostname      whoami::hostname():    rpi4g
Device's Platform      whoami::platform():    BSD
Device's OS Distro     whoami::distro():      FreeBSD 13.0-CURRENT
Device's Desktop Env.  whoami::desktop_env(): Unknown: Unknown
freebsd@rpi4g ~/c/whoami (master)> cargo r --release --example whoami-demo
    Finished release [optimized] target(s) in 0.07s
     Running `target/release/examples/whoami-demo`
WhoAmI 1.0.0

User's Name            whoami::realname():    Freebsd
[src/unix.rs:157] string_from_os(username_os()) = "freebsd"
fish: 'cargo r --release --example who…' terminated by signal SIGBUS (Misaligned address error)
```

**Comment by AldaronLau** [comment]
@fundon Does A:

```rust
fn main() {
    println!("WhoAmI {}", env!("CARGO_PKG_VERSION"));
    println!();
    println!(
        "User's Name            whoami::realname():    {}",
        whoami::realname()
    );
    println!(
        "User's Username        whoami::username():    {}",
        whoami::username()
    );
}
```

crash,  or B:

```rust
fn main() {
    println!("WhoAmI {}", env!("CARGO_PKG_VERSION"));
    println!();
    println!(
        "User's Name            whoami::realname():    {}",
        whoami::realname()
    );
}
```

crash, or C:

```rust
fn main() {
    println!(
        "User's Username        whoami::username():    {}",
        whoami::username()
    );
}
```

crash?

**Comment by fundon** [comment]
C crash
```rust
whoami::username()
```

**Comment by AldaronLau** [comment]
@fundon How about if you change it to `whoami::username_os`?

The inlined function should look something like (if you want to see if this crashes as well):
```rust
extern "system" {
    fn strlen(cs: *const c_void) -> usize;
}

fn whoami_username_os() -> OsString {
    const BUF_SIZE: usize = 16_384; // size from the man page
    let mut buffer = mem::MaybeUninit::<[u8; BUF_SIZE]>::uninit();
    let mut passwd = mem::MaybeUninit::<PassWd>::uninit();
    let mut _passwd = mem::MaybeUninit::<*mut PassWd>::uninit();

    // Get PassWd `struct`.
    let passwd = unsafe {
        getpwuid_r(
            geteuid(),
            passwd.as_mut_ptr(),
            buffer.as_mut_ptr() as *mut c_void,
            BUF_SIZE,
            _passwd.as_mut_ptr(),
        );

        passwd.assume_init()
    };

    if passwd.pw_name.is_null() {
        return "".to_string().into();
    }

    // Get a byte slice of the c string.
    let slice = unsafe {
        let length = strlen(passwd.pw_name);
        std::slice::from_raw_parts(passwd.pw_name as *const u8, length)
    };

    // Turn byte slice into Rust String.
    OsString::from_vec(slice.to_vec())
}
```

**Comment by AldaronLau** [comment]
@fundon I just figured out what's going on.  It's a bug in whoami, where I need to replace:

```rust
#[repr(C)]
struct PassWd {
    pw_name: *const c_void,
    pw_passwd: *const c_void,
    pw_uid: u32,
    pw_gid: u32,
    #[cfg(target_os = "macos")]
    pw_change: isize,
    #[cfg(target_os = "macos")]
    pw_class: *const c_void,
    pw_gecos: *const c_void,
    pw_dir: *const c_void,
    pw_shell: *const c_void,
    #[cfg(target_os = "macos")]
    pw_expire: isize,
    #[cfg(target_os = "macos")]
    pw_fields: i32,
}
```

with

```rust
#[repr(C)]
struct PassWd {
    pw_name: *const c_void,
    pw_passwd: *const c_void,
    pw_uid: u32,
    pw_gid: u32,
    #[cfg(any(target_os = "macos", target_os = "freebsd"))]
    pw_change: isize,
    #[cfg(any(target_os = "macos", target_os = "freebsd"))]
    pw_class: *const c_void,
    pw_gecos: *const c_void,
    pw_dir: *const c_void,
    pw_shell: *const c_void,
    #[cfg(any(target_os = "macos", target_os = "freebsd"))]
    pw_expire: isize,
    #[cfg(any(target_os = "macos", target_os = "freebsd"))]
    pw_fields: i32,
}
```

Since the definitions are the same for MacOS and FreeBSD.  I'll release a patched version of whoami soon with the bugfix.

**Comment by fundon** [comment]
And add those:
```
-#[cfg(not(target_os = "macos"))]
+#[cfg(not(any(target_os = "macos", target_os = "freebsd")))]
 #[inline(always)]
 pub const fn platform() -> Platform {
     Platform::Linux
 }
+
+#[cfg(target_os = "freebsd")]
+pub const fn platform() -> Platform {
+    Platform::Bsd
+}
```

**Comment by fundon** [comment]
@AldaronLau I tested your `PassWd` patched. It has fixed it.

**Comment by AldaronLau** [comment]
@fundon Perfect!

**Comment by AldaronLau** [comment]
@mehcode @fundon This has been fixed in whoami v1.0.1.  Which means for sqlx to support FreeBSD, it needs to upgrade to whoami 1.0.  This should be really quick to do, as it almost certainly won't break the sqlx code.

**Comment by mehcode** [comment]
Thanks for getting this investigated and resolved! 🎉
